autograder-core/src/main/java/de/firemage/autograder/core/check/complexity/RedundantConstructorCheck.java

```diff
@@ -2,14 +2,62 @@
 
 import de.firemage.autograder.core.LocalizedMessage;
 import de.firemage.autograder.core.ProblemType;
+import de.firemage.autograder.core.Translatable;
 import de.firemage.autograder.core.check.ExecutableCheck;
-import de.firemage.autograder.core.pmd.PMDCheck;
-import net.sourceforge.pmd.lang.java.rule.codestyle.UnnecessaryConstructorRule;
+import de.firemage.autograder.core.dynamic.DynamicAnalysis;
+import de.firemage.autograder.core.integrated.IntegratedCheck;
+import de.firemage.autograder.core.integrated.StaticAnalysis;
+import spoon.processing.AbstractProcessor;
+import spoon.reflect.code.CtBlock;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtConstructor;
+import spoon.reflect.declaration.CtElement;
+
+import java.util.function.Predicate;
 
 @ExecutableCheck(reportedProblems = {ProblemType.REDUNDANT_DEFAULT_CONSTRUCTOR})
-public class RedundantConstructorCheck extends PMDCheck {
-    public RedundantConstructorCheck() {
-        super(new LocalizedMessage("implicit-constructor-exp"),
-            new UnnecessaryConstructorRule(), ProblemType.REDUNDANT_DEFAULT_CONSTRUCTOR);
+public class RedundantConstructorCheck extends IntegratedCheck {
+    private static final Translatable MESSAGE = new LocalizedMessage("implicit-constructor-exp");
+
+    @Override
+    protected void check(StaticAnalysis staticAnalysis, DynamicAnalysis dynamicAnalysis) {
+        staticAnalysis.processWith(new AbstractProcessor<CtClass<?>>() {
+            @Override
+            public void process(CtClass<?> element) {
+                var ctors = element.getConstructors();
+                if (ctors.size() != 1) return;
+
+                var ctor = ctors.iterator().next();
+                var isRedundant = !ctor.isImplicit()
+                    && ctor.getParameters().isEmpty()
+                    && hasDefaultVisibility(element, ctor)
+                    && isEmpty(ctor.getBody())
+                    && ctor.getThrownTypes().isEmpty();
+                if (isRedundant) {
+                    addLocalProblem(ctor, MESSAGE, ProblemType.REDUNDANT_DEFAULT_CONSTRUCTOR);
+                }
+            }
+        });
+    }
+
+    private boolean hasDefaultVisibility(CtClass<?> type, CtConstructor<?> ctor) {
```

Comment:
Why default visibility? The java generated default constructor is public

Reply:
https://docs.oracle.com/javase/specs/jls/se22/html/jls-8.html#jls-8.8.9

> The default constructor has the same access modifier as the class, unless the class lacks an access modifier, in which case the default constructor has package access